crates/json/src/schema/index.rs

```diff
@@ -54,23 +56,61 @@ where
     }
 
     pub fn verify_references(&self) -> Result<(), Error> {
-        for (&curi, &schema) in &self.0 {
-            for kw in &schema.kw {
-                if let Keyword::Application(Application::Ref(ruri), _) = kw {
-                    if !self.0.contains_key(ruri.as_str()) {
-                        return Err(Error::InvalidReference {
-                            ruri: ruri.clone(),
-                            curi: curi.to_owned(),
-                        });
-                    }
-                }
+        for (referrer, referrent) in self.references() {
+            if !self.0.contains_key(referrent) {
+                return Err(Error::InvalidReference {
+                    ruri: referrent.clone(),
+                    curi: referrer.clone(),
+                });
             }
         }
         Ok(())
     }
 
+    pub fn into_index(self) -> Index<'s, A> {
+        let referrents: BTreeSet<_> = self.references().map(|(_, r)| r).collect();
+
+        let (fast, slow) = self.0.into_iter().partition(|e| referrents.contains(&e.0));
+        Index { fast, slow }
+    }
+
+    fn references<'a>(&'a self) -> impl Iterator<Item = (&'s url::Url, &'s url::Url)> + 'a {
+        self.0.iter().flat_map(|(referrer, schema)| {
+            schema.kw.iter().filter_map(move |kw| match kw {
+                Keyword::Application(Application::Ref(referrent), _) => {
+                    Some((*referrer, referrent))
+                }
+                _ => None,
+            })
+        })
+    }
+}
+
+/// Index is a packed, sorted index over Schema references.
+/// It provides lookups over Schema canonical and anchor-form URIs.
+pub struct Index<'s, A>
+where
+    A: Annotation,
+{
+    // Store is subdivided between a |fast| and |slow| index.
+    // |fast| items are statically known to be referenced, and there are fewer of them.
+    // |slow| items may still be referenced, and there are more of them.
```

Comment:
Do you have any anecdotal, order-of-magnitude guess at how many items are ending up in `fast` vs `slow` in the workloads you're trying out? I think my intuition about this must be wrong.

Reply:
maybe 4 vs 50 or so in my workload? though heavily dependent on the catalog of course.

Basically every single location in a schema file is a schema, with a canonical URL, which _could_ be referenced. That's going into `slow`. Only schemas which appear as a `$ref `of something else go into `fast`.